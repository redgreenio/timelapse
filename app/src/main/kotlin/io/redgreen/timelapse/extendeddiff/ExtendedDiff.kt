package io.redgreen.timelapse.extendeddiff

sealed class ExtendedDiff {
  data class NoChanges(val sourceCode: String) : ExtendedDiff()

  data class HasChanges(
    val sourceCode: String,
    val comparisonResults: List<ComparisonResult>
  ) : ExtendedDiff()
}
