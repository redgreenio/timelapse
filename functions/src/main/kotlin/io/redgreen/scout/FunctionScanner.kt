package io.redgreen.scout

interface FunctionScanner {
  fun scan(snippet: String): List<PossibleFunction>
}
