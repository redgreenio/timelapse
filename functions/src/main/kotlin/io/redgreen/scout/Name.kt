package io.redgreen.scout

data class Name(val value: String)
