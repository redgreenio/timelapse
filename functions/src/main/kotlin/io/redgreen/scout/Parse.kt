package io.redgreen.scout

import io.redgreen.scout.ScanMode.FIND_BLOCKS
import io.redgreen.scout.ScanMode.SKIP_MULTILINE_COMMENT
import io.redgreen.scout.ScanMode.SKIP_MULTILINE_STRING_LITERAL
import io.redgreen.scout.ScanMode.SKIP_SINGLE_LINE_COMMENT
import io.redgreen.scout.ScanMode.SKIP_STRING_LITERAL
import io.redgreen.scout.extensions.contents
import io.redgreen.scout.extensions.push
import io.redgreen.scout.extensions.top
import java.util.Stack

typealias Depth = Int

private enum class ScanMode {
  FIND_BLOCKS,
  SKIP_SINGLE_LINE_COMMENT,
  SKIP_MULTILINE_COMMENT,
  SKIP_STRING_LITERAL,
  SKIP_MULTILINE_STRING_LITERAL,
}

private const val TOKEN_OPEN_CURLY = '{'
private const val TOKEN_CLOSE_CURLY = '}'
private const val TOKEN_FORWARD_SLASH = '/'
private const val TOKEN_NEW_LINE = '\n'
private const val TOKEN_ASTERISK = '*'
private const val TOKEN_STRING_QUOTE = '"'
private const val TOKEN_MULTILINE_STRING = "\"\"\""

private const val PREVIOUS_CHARS_BUFFER_SIZE = 2

@SuppressWarnings("LoopWithTooManyJumpStatements", "LongMethod", "ComplexMethod", "NestedBlockDepth", "ReturnCount")
fun parse(snippet: String, lineNumberOffset: Int = 0): ParseResult {
  val snippetChars = snippet.toCharArray()
  val previousCharsBuffer = CharArray(PREVIOUS_CHARS_BUFFER_SIZE)

  var mode = FIND_BLOCKS
  var maximumDepth = 0
  var lineNumber = if (lineNumberOffset == 0) 1 else lineNumberOffset
  var startLineNumber = lineNumber
  val depthStack = Stack<Depth>()
  for (char in snippetChars) {
    if (char == TOKEN_NEW_LINE) {
      lineNumber++
    }

    val quoteFoundWhenLookingForBlocks = char == TOKEN_STRING_QUOTE && mode == FIND_BLOCKS
    if (quoteFoundWhenLookingForBlocks && isMultilineString(previousCharsBuffer, char)) {
      mode = SKIP_MULTILINE_STRING_LITERAL
      previousCharsBuffer.push(char)
      continue
    } else if (quoteFoundWhenLookingForBlocks) {
      mode = SKIP_STRING_LITERAL
      previousCharsBuffer.push(char)
      continue
    }

    when {
      isSingleLineComment(mode, previousCharsBuffer, char) -> mode = SKIP_SINGLE_LINE_COMMENT
      isMultilineComment(mode, previousCharsBuffer, char) -> mode = SKIP_MULTILINE_COMMENT
    }

    when (mode) {
      SKIP_STRING_LITERAL -> if (char == TOKEN_STRING_QUOTE) {
        mode = FIND_BLOCKS
      }

      SKIP_SINGLE_LINE_COMMENT -> if (char == TOKEN_NEW_LINE) {
        mode = FIND_BLOCKS
      }

      SKIP_MULTILINE_COMMENT -> if (previousCharsBuffer.top().get() == TOKEN_ASTERISK && char == TOKEN_FORWARD_SLASH) {
        mode = FIND_BLOCKS
      }

      FIND_BLOCKS -> when (char) {
        TOKEN_OPEN_CURLY -> {
          if (depthStack.isEmpty()) {
            maximumDepth = 1
            startLineNumber = lineNumber
            depthStack.push(maximumDepth)
          } else {
            val depth = depthStack.peek() + 1
            if (depth > maximumDepth) {
              maximumDepth = depth
            }
            depthStack.push(depth)
          }
        }

        TOKEN_CLOSE_CURLY -> {
          if (depthStack.isEmpty()) {
            return ParseResult.malformedFunction(startLineNumber, lineNumber)
          }

          depthStack.pop()

          if (depthStack.isEmpty()) {
            return ParseResult.wellFormedFunction(startLineNumber, lineNumber, maximumDepth)
          }
        }
      }

      SKIP_MULTILINE_STRING_LITERAL -> if (isMultilineString(previousCharsBuffer, char)) {
        mode = FIND_BLOCKS
      }
    }

    previousCharsBuffer.push(char)
  }

  return ParseResult.nothing(startLineNumber, lineNumber)
}

private fun isSingleLineComment(
  currentScanMode: ScanMode,
  previousCharsBuffer: CharArray,
  char: Char
): Boolean {
  return previousCharsBuffer.top().isPresent &&
    previousCharsBuffer.top().get() == TOKEN_FORWARD_SLASH &&
    char == TOKEN_FORWARD_SLASH &&
    currentScanMode != SKIP_STRING_LITERAL
}

private fun isMultilineComment(
  currentScanMode: ScanMode,
  previousCharsBuffer: CharArray,
  char: Char
): Boolean {
  return previousCharsBuffer.top().isPresent &&
    previousCharsBuffer.top().get() == TOKEN_FORWARD_SLASH &&
    char == TOKEN_ASTERISK &&
    currentScanMode != SKIP_SINGLE_LINE_COMMENT
}

private fun isMultilineString(
  previousCharsBuffer: CharArray,
  char: Char
): Boolean {
  val bufferContents = previousCharsBuffer.contents()
  return bufferContents.isPresent &&
    "$bufferContents$char" == TOKEN_MULTILINE_STRING
}
