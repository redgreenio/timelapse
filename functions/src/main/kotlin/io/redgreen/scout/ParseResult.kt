package io.redgreen.scout

import java.util.Optional

sealed class ParseResult(
  open val startLine: Int,
  open val endLine: Int
) {
  val length: Int
    get() = endLine - startLine + 1

  data class Nothing(
    override val startLine: Int,
    override val endLine: Int
  ) : ParseResult(startLine, endLine)

  data class WellFormedFunction(
    val name: Optional<Name> = Optional.empty(),
    override val startLine: Int,
    override val endLine: Int,
    val depth: Depth
  ) : ParseResult(startLine, endLine) {
    fun addName(name: String): ParseResult {
      return copy(name = Optional.of(Name(name)))
    }
  }

  data class MalformedFunction(
    override val startLine: Int,
    override val endLine: Int
  ) : ParseResult(startLine, endLine)

  companion object {
    fun nothing(startLine: Int, endLine: Int): Nothing {
      check(startLine, endLine)
      return Nothing(startLine, endLine)
    }

    fun wellFormedFunction(startLine: Int, endLine: Int, depth: Int): WellFormedFunction {
      return wellFormedFunction(Optional.empty(), startLine, endLine, depth)
    }

    fun wellFormedFunction(name: String, startLine: Int, endLine: Int, depth: Int): WellFormedFunction {
      return wellFormedFunction(Optional.of(Name(name)), startLine, endLine, depth)
    }

    private fun wellFormedFunction(name: Optional<Name>, startLine: Int, endLine: Int, depth: Int): WellFormedFunction {
      check(startLine, endLine)
      check(depth >= 0) { "`depth`: $depth should be a positive integer" }
      check(!(startLine == 0 && endLine == 0 && depth != 0)) { "`depth` must be zero for a non-existent function, but was `$depth`" }
      return WellFormedFunction(name, startLine, endLine, depth)
    }

    fun malformedFunction(startLine: Int, endLine: Int): MalformedFunction {
      check(startLine, endLine)
      return MalformedFunction(startLine, endLine)
    }

    private fun check(startLine: Int, endLine: Int) {
      check(startLine >= 0) { "`startLine`: $startLine should be a positive integer" }
      check(endLine >= 0) { "`endLine`: $endLine should be a positive integer" }
      check(startLine <= endLine) { "`startLine`: $startLine cannot be greater than `endLine`: $endLine" }
    }
  }
}
