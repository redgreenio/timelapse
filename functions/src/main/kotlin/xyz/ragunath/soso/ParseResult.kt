package xyz.ragunath.soso

sealed class ParseResult(
  open val startLine: Int,
  open val endLine: Int
) {
  val length: Int
    get() = endLine - startLine + 1

  companion object {
    fun nothing(startLine: Int, endLine: Int): Nothing {
      check(startLine, endLine)
      return Nothing(startLine, endLine)
    }

    fun wellFormedFunction(startLine: Int, endLine: Int, depth: Int): WellFormedFunction {
      check(startLine, endLine)
      check(depth >= 0) { "`depth`: $depth should be a positive integer" }
      check(!(startLine == 0 && endLine == 0 && depth != 0)) { "`depth` must be zero for a non-existent function, but was `$depth`" }
      return WellFormedFunction(startLine, endLine, depth)
    }

    fun malformedFunction(startLine: Int, endLine: Int): MalformedFunction {
      check(startLine, endLine)
      return MalformedFunction(startLine, endLine)
    }

    private fun check(startLine: Int, endLine: Int) {
      check(startLine >= 0) { "`startLine`: $startLine should be a positive integer" }
      check(endLine >= 0) { "`endLine`: $endLine should be a positive integer" }
      check(startLine <= endLine) { "`startLine`: $startLine cannot be greater than `endLine`: $endLine" }
    }
  }

  data class Nothing(
    override val startLine: Int,
    override val endLine: Int
  ) : ParseResult(startLine, endLine)

  data class WellFormedFunction(
    override val startLine: Int,
    override val endLine: Int,
    val depth: Depth
  ) : ParseResult(startLine, endLine)

  data class MalformedFunction(
    override val startLine: Int,
    override val endLine: Int
  ) : ParseResult(startLine, endLine)
}
