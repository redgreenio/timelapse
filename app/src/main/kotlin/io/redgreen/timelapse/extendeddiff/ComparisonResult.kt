package io.redgreen.timelapse.extendeddiff

import io.redgreen.scout.ParseResult.WellFormedFunction

sealed class ComparisonResult {
  data class Added(val function: WellFormedFunction) : ComparisonResult()
  data class Deleted(val function: WellFormedFunction, val snippet: String) : ComparisonResult()
  data class Modified(val function: WellFormedFunction) : ComparisonResult()
}
