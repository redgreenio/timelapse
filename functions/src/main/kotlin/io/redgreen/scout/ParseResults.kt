package io.redgreen.scout

import io.redgreen.scout.ParseResult.WellFormedFunction

fun getParseResults(
  scanner: (String) -> List<PossibleFunction>,
  snippet: String
): List<ParseResult> {
  val possibleFunctions = scanner(snippet)
  if (possibleFunctions.isEmpty()) {
    return emptyList()
  }

  val lineNumbers = possibleFunctions.map { it.startLineNumber }
  val functionSnippets = split(snippet, lineNumbers.first(), *lineNumbers.drop(1).toIntArray())
  val parseResults = functionSnippets
    .zip(lineNumbers)
    .map { (functionSnippet, lineNumber) -> parse(functionSnippet, lineNumber) }

  return possibleFunctions.zip(parseResults) { possibleFunction, result ->
    if (result is WellFormedFunction) {
      result.addName(possibleFunction.name)
    } else {
      result
    }
  }
}
