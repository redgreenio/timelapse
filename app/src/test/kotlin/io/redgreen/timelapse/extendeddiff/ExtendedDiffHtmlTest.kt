package io.redgreen.timelapse.extendeddiff

import io.redgreen.scout.ParseResult
import io.redgreen.timelapse.extendeddiff.ComparisonResult.Added
import io.redgreen.timelapse.extendeddiff.ComparisonResult.Deleted
import io.redgreen.timelapse.extendeddiff.ComparisonResult.Modified
import io.redgreen.timelapse.extendeddiff.ComparisonResult.Renamed
import io.redgreen.timelapse.extendeddiff.ComparisonResult.Unmodified
import io.redgreen.timelapse.extendeddiff.ExtendedDiff.HasChanges
import io.redgreen.timelapse.extendeddiff.ExtendedDiff.NoChanges
import org.approvaltests.Approvals
import org.approvaltests.reporters.QuietReporter
import org.approvaltests.reporters.UseReporter
import org.junit.jupiter.api.Test

@UseReporter(QuietReporter::class)
class ExtendedDiffHtmlTest {
  @Test
  fun `it should handle added function`() {
    // given
    val kotlinSource = """
      fun a() {}
      fun b() {}
    """.trimIndent()

    val comparisonResults = listOf(
      Added(ParseResult.wellFormedFunction("b", 2, 2, 1))
    )

    val hasChanges = HasChanges(kotlinSource, comparisonResults)

    // when & then
    Approvals.verifyHtml(hasChanges.toHtml())
  }

  @Test
  fun `it should handle modified function`() {
    // given
    val swiftSource = """
      class SomeClass {
        func a() {
          // Hello, world!
        }
      }
    """.trimIndent()

    val snippetA = """
      |  func a() {
      |    // Hello, world!
      |  }
    """.trimMargin("|")
    val comparisonResults = listOf(
      Modified(ParseResult.wellFormedFunction("a", 2, 4, 1), snippetA)
    )

    val hasChanges = HasChanges(swiftSource, comparisonResults)

    // when & then
    Approvals.verifyHtml(hasChanges.toHtml())
  }

  @Test
  fun `it should handle deleted function`() {
    // given
    val kotlinSource = """
      fun b() {
      }
    """.trimIndent()

    val snippetA = """
      fun a() {
      }
    """.trimIndent()

    val comparisonResults = listOf(
      Deleted(ParseResult.wellFormedFunction("a", 1, 2, 1), snippetA)
    )

    val hasChanges = HasChanges(kotlinSource, comparisonResults)

    // when & then
    Approvals.verifyHtml(hasChanges.toHtml())
  }

  @Test
  fun `it should handle modified and added functions`() {
    // given
    val swiftSource = """
      class SomeClass {
        func a() {
          // Hello, world!
        }
        
        fun b() {
          println("I am new!")
        }
      }
    """.trimIndent()

    val snippetA = """
      |  func a() {
      |    // Hello, world!
      |  }
    """.trimIndent()
    val comparisonResults = listOf(
      Modified(ParseResult.wellFormedFunction("a", 2, 4, 1), snippetA),
      Added(ParseResult.wellFormedFunction("b", 6, 8, 1)),
    )

    val hasChanges = HasChanges(swiftSource, comparisonResults)

    // when & then
    Approvals.verifyHtml(hasChanges.toHtml())
  }

  @Test
  fun `it should handle added, modified and deleted functions`() {
    // given
    val kotlinSource = """
      function b() {
      }
      
      
      function d() {
        println("Knock, knock!")
      }
      
    """.trimIndent()

    val functionA = """
      fun a() {
        // Nothing here...
      }
    """.trimIndent()
    val functionB = """
      function b() {
      }
    """.trimIndent()
    val functionC = """
      fun c(a: Int, b: Int): Int {
        return a + b
      }
    """.trimIndent()
    val functionE = """
      fun e(a: Double, b: Double): Double {
        return a * b
      }
    """.trimIndent()

    val comparisonResults = listOf(
      Deleted(ParseResult.wellFormedFunction("a", 1, 3, 1), functionA),
      Modified(ParseResult.wellFormedFunction("b", 1, 2, 1), functionB),
      Deleted(ParseResult.wellFormedFunction("c", 4, 6, 1), functionC),
      Added(ParseResult.wellFormedFunction("d", 5, 7, 1)),
      Deleted(ParseResult.wellFormedFunction("e", 8, 10, 1), functionE)
    )

    // when
    val hasChanges = HasChanges(kotlinSource, comparisonResults)

    // then
    Approvals.verifyHtml(hasChanges.toHtml())
  }

  @Test
  fun `deleted functions should not overlap added functions`() {
    // given
    val kotlinSource = """
      fun newlyAddedFunction() {
        println("Hello, world!")
      }
    """.trimIndent()

    val deletedFunction = """
      fun deletedFunction() {
      }
    """.trimIndent()

    val comparisonResults = listOf(
      Added(ParseResult.wellFormedFunction("newlyAddedFunction", 1, 3, 1)),
      Deleted(ParseResult.wellFormedFunction("deletedFunction", 2, 3, 1), deletedFunction),
    )

    // when
    val hasChanges = HasChanges(kotlinSource, comparisonResults)

    // then
    Approvals.verifyHtml(hasChanges.toHtml())
  }

  @Test
  fun `deleted functions should not overlap modified functions`() {
    // given
    val kotlinSource = """
      fun modifiedFunction() {
        println("Hello, world!")
      }
    """.trimIndent()

    val deletedFunction = """
      fun deletedFunction() {
      }
    """.trimIndent()

    val comparisonResults = listOf(
      Added(ParseResult.wellFormedFunction("modifiedFunction", 1, 3, 1)),
      Deleted(ParseResult.wellFormedFunction("deletedFunction", 2, 3, 1), deletedFunction),
    )

    // when
    val hasChanges = HasChanges(kotlinSource, comparisonResults)

    // then
    Approvals.verifyHtml(hasChanges.toHtml())
  }

  @Test
  fun `it should handle comparison results in any order`() {
    // given
    val kotlinSource = """
      function b() {
      }
      
      
      function d() {
        println("Knock, knock!")
      }
      
    """.trimIndent()

    val functionA = """
      fun a() {
        // Nothing here...
      }
    """.trimIndent()
    val functionB = """
      function b() {
      }
    """.trimIndent()
    val functionC = """
      fun c(a: Int, b: Int): Int {
        return a + b
      }
    """.trimIndent()
    val functionE = """
      fun e(a: Double, b: Double): Double {
        return a * b
      }
    """.trimIndent()

    val comparisonResults = listOf(
      Deleted(ParseResult.wellFormedFunction("e", 8, 10, 1), functionE),
      Added(ParseResult.wellFormedFunction("d", 5, 7, 1)),
      Deleted(ParseResult.wellFormedFunction("c", 4, 6, 1), functionC),
      Deleted(ParseResult.wellFormedFunction("a", 1, 3, 1), functionA),
      Modified(ParseResult.wellFormedFunction("b", 1, 2, 1), functionB)
    )

    // when
    val hasChanges = HasChanges(kotlinSource, comparisonResults)

    // then
    Approvals.verifyHtml(hasChanges.toHtml())
  }

  @Test
  fun `it should render HTML for no changes`() {
    // given
    val kotlinSource = """
      fun sayHello() {
        println("Hey, hey!")
      }
    """.trimIndent()
    val noChanges = NoChanges(kotlinSource)

    // when & then
    Approvals.verifyHtml(noChanges.toHtml())
  }

  @Test
  fun `it should not overlap deleted functions in-between unchanged functions`() {
    // given
    val kotlinSource = """
      fun getReview(book: Book): Review {
        val rating = book.read()
        val ratingFromFriends = book.discussWithFriends()
        return Review(rating + ratingFromFriends)
      }
    """.trimIndent()

    val getAuthorFunction = """
      fun getAuthor(book: Book): Author {
        return book.author
      }
    """.trimIndent()

    val comparisonResults = listOf(
      Deleted(ParseResult.wellFormedFunction("getAuthor", 2, 4, 1), getAuthorFunction),
      Unmodified(ParseResult.wellFormedFunction("getReview", 1, 5, 1)),
    )
    val hasChanges = HasChanges(kotlinSource, comparisonResults)

    // when & then
    Approvals.verifyHtml(hasChanges.toHtml())
  }

  @Test
  fun `it should render renamed function`() {
    // given
    val kotlinSource = """
      fun add(x: Int, y: Int): Int {
        return x + y
      }
    """.trimIndent()

    val comparisonResults = listOf(
      Renamed(ParseResult.wellFormedFunction("add", 1, 3, 1), "plus")
    )
    val hasChanges = HasChanges(kotlinSource, comparisonResults)

    // when & then
    Approvals.verifyHtml(hasChanges.toHtml())
  }
}
