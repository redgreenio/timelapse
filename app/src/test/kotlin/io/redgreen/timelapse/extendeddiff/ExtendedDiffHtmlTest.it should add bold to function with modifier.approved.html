<html lang="en-US">
<head>
  <style>
    .added {
      background-color: #e6ffed;
    }
    .modified {
      background-color: #dbedff80;
    }
    .deleted {
      background-color: #ffdce0;
    }
    .readability {
      background-color: #fafafa;
    }
    table {
      width: 100%;
      border-collapse: collapse;
    }
    table td {
      font-family: "monospace";
      font-size: small;
    }
  </style>
</head>
<body>
<table>
  <tbody>
    <tr class="modified">
      <td>1</td>
      <td>private inline fun <b>greet</b>() = println(&quot;Hello, world&quot;)</td>
    </tr>
  </tbody>
</table>
</body>
</html>
